### PR TITLE
add microbenchmark for exchange and window function

### LINKED ISSUE #4276: Can't build tiflash within exchange_perftest tests.
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

1. cmake .. -DCMAKE_BUILD_TYPE=Debug -DENABLE_TESTS=ON
2. make -j

Then will got error
```
/xxxxxxx/dbms/src/Flash/Mpp/MPPTunnel.cpp:67:17: error: no member named 'tryFlushOne' in 'DB::tests::(anonymous namespace)::MockWriter'
        writer->tryFlushOne();
        ~~~~~~  ^
/xxxxxxx/dbms/src/Flash/Mpp/MPPTunnel.cpp:51:13: note: in instantiation of member function 'DB::MPPTunnelBase<DB::tests::(anonymous namespace)::MockWriter>::finishSendQueue' requested here
            finishSendQueue();
            ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.3.sdk/usr/include/c++/v1/memory:1933:3: note: in instantiation of member function 'DB::MPPTunnelBase<DB::tests::(anonymous namespace)::MockWriter>::~MPPTunnelBase' requested here
  __compressed_pair_elem(piecewise_construct_t, tuple<_Args...> __args,
  ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.3.sdk/usr/include/c++/v1/memory:4038:26: note: in instantiation of function template specialization 'std::__1::__shared_ptr_emplace<DB::MPPTunnelBase<DB::tests::(anonymous namespace)::MockWriter>, std::__1::allocator<DB::MPPTunnelBase<DB::tests::(anonymous namespace)::MockWriter>>>::__shared_ptr_emplace<mpp::TaskMeta &, mpp::TaskMeta &, std::__1::chrono::duration<long long>, const int &, bool>' requested here
    ::new(__hold2.get()) _CntrlBlk(__a2, _VSTD::forward<_Args>(__args)...);
                         ^
/xxxxxxx/dbms/src/Flash/tests/exchange_perftest.cpp:505:32: note: in instantiation of function template specialization 'std::__1::make_shared<DB::MPPTunnelBase<DB::tests::(anonymous namespace)::MockWriter>, mpp::TaskMeta &, mpp::TaskMeta &, std::__1::chrono::duration<long long>, const int &, bool>' requested here
            auto tunnel = std::make_shared<MockTunnel>(
                               ^
In file included from /xxxxxxx/dbms/src/Flash/tests/exchange_perftest.cpp:15:
/xxxxxxx/dbms/src/Flash/Mpp/MPPTunnel.cpp:151:26: error: no member named 'write' in 'DB::tests::(anonymous namespace)::MockWriter'
            if (!writer->write(*res))
                 ~~~~~~  ^
/xxxxxxx/dbms/src/Flash/Mpp/MPPTunnel.cpp:235:21: note: in instantiation of member function 'DB::MPPTunnelBase<DB::tests::(anonymous namespace)::MockWriter>::sendJob' requested here
                    sendJob();
                    ^
/xxxxxxx/dbms/src/Flash/tests/exchange_perftest.cpp:511:21: note: in instantiation of member function 'DB::MPPTunnelBase<DB::tests::(anonymous namespace)::MockWriter>::connect' requested here
            tunnel->connect(writer.get());
                    ^
In file included from /xxxxxxx/dbms/src/Flash/tests/exchange_perftest.cpp:15:
/xxxxxxx/dbms/src/Flash/Mpp/MPPTunnel.cpp:179:17: error: no member named 'writeDone' in 'DB::tests::(anonymous namespace)::MockWriter'
        writer->writeDone(grpc::Status::OK);
        ~~~~~~  ^
/xxxxxxx/dbms/src/Flash/Mpp/MPPTunnel.cpp:122:25: error: no member named 'tryFlushOne' in 'DB::tests::(anonymous namespace)::MockWriter'
                writer->tryFlushOne();
                ~~~~~~  ^
/xxxxxxx/dbms/src/Flash/Mpp/MPPTunnelSet.cpp:94:28: note: in instantiation of member function 'DB::MPPTunnelBase<DB::tests::(anonymous namespace)::MockWriter>::write' requested here
    tunnels[partition_id]->write(packet);
                           ^
/xxxxxxx/dbms/src/Flash/Coprocessor/StreamingDAGResponseWriter.cpp:282:25: note: in instantiation of member function 'DB::MPPTunnelSetBase<DB::MPPTunnelBase<DB::tests::(anonymous namespace)::MockWriter>>::write' requested here
                writer->write(packet[part_id], part_id);
                        ^
/xxxxxxx/dbms/src/Flash/Coprocessor/StreamingDAGResponseWriter.cpp:296:9: note: in instantiation of function template specialization 'DB::StreamingDAGResponseWriter<std::__1::shared_ptr<DB::MPPTunnelSetBase<DB::MPPTunnelBase<DB::tests::(anonymous namespace)::MockWriter>>>>::partitionAndEncodeThenWriteBlocks<false>' requested here
        partitionAndEncodeThenWriteBlocks<send_exec_summary_at_last>(blocks, response);
        ^
/xxxxxxx/dbms/src/Flash/Coprocessor/StreamingDAGResponseWriter.cpp:82:9: note: in instantiation of function template specialization 'DB::StreamingDAGResponseWriter<std::__1::shared_ptr<DB::MPPTunnelSetBase<DB::MPPTunnelBase<DB::tests::(anonymous namespace)::MockWriter>>>>::batchWrite<false>' requested here
        batchWrite<false>();
        ^
/xxxxxxx/dbms/src/Flash/Coprocessor/StreamingDAGResponseWriter.h:26:5: note: in instantiation of member function 'DB::StreamingDAGResponseWriter<std::__1::shared_ptr<DB::MPPTunnelSetBase<DB::MPPTunnelBase<DB::tests::(anonymous namespace)::MockWriter>>>>::write' requested here
    StreamingDAGResponseWriter(
    ^
/xxxxxxx/dbms/src/Flash/tests/exchange_perftest.cpp:537:21: note: in instantiation of member function 'DB::StreamingDAGResponseWriter<std::__1::shared_ptr<DB::MPPTunnelSetBase<DB::MPPTunnelBase<DB::tests::(anonymous namespace)::MockWriter>>>>::StreamingDAGResponseWriter' requested here
                new StreamingDAGResponseWriter<MockTunnelSetPtr>(
                    ^
In file included from /xxxxxxx/dbms/src/Flash/tests/exchange_perftest.cpp:13:
/xxxxxxx/dbms/src/Flash/Coprocessor/StreamingDAGResponseWriter.cpp:105:29: error: no matching member function for call to 'write'
                    writer->write(packet);
                    ~~~~~~~~^~~~~
```

I think this build issue come from #4120

### LINKED ISSUE #5138: add microbenchmark for exchange and window function
## Enhancement
1. Fine grained shuffle need some microbenchmark to measure performance.
2. We already had [exchange_perftest.cpp](https://github.com/pingcap/tiflash/blob/master/dbms/src/Flash/tests/exchange_perftest.cpp), we can use google benchmark to refine it.

### PR DESCRIPTION
**Pull request by guo-shaoge**
Signed-off-by: guo-shaoge <shaoge1994@163.com>

### What problem does this PR solve?

Issue Number: close #5138, close #4276

Problem Summary: 
1. To help compare performance of fine grained shuffle, add exchange and window function microbenchmark.
2. https://github.com/pingcap/tiflash/pull/5048 will use this benchmark as baseline.

### What is changed and how it works?
1. bench_exchange.cpp
  3.1. sender: unionBlockInputStream <- ExchangeSenderBlockInputStream(x concurrency)<- MockFixedRowBlockInputStream(x concurrency)
  2.2. receiver: unionBlockInputStream <- ExchangeReceiverBlockInputStream(x concurrency)<- ExchangeReceiver
4. bench_window.cpp:
  3.1. sender:  unionBlockInputStream <- WindowBlockInputStream <- MergeSortingBlockInputStream(x concurrency) <- PartialSortInputStream(x concurrency) <- ExchangeSenderBlockInputStream(x concurrency)<- MockFixedRowBlockInputStream(x concurrency)
  3.2. receiver: unionBlockInputStream <- ExchangeReceiverBlockInputStream(x concurrency)<- ExchangeReceiver
5. exchange_perftest.cpp: move code to bench_exchange.cpp

Usage:
1. cmake .. -DCMAKE_BUILD_TYPE=Release -DENABLE_TESTS=ON && make bench_dbms -j15
6. cd build && ./dbms/bench_dbms --benchmark_filter='WindowFunctionBench.*'
7. cd build && ./dbms/bench_dbms --benchmark_filter='ExchangeBench.*'

Result:
image
image

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by ywqzzy** [comment]
This pr is also related to https://github.com/pingcap/tiflash/issues/4276

File: dbms/src/Flash/tests/bench_window.cpp [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
...ng sort_json = R"({"byItems":[{"expr":{"tp":"ColumnRef","val":"gAAAAAAAAAA=","sig":"Unspecified","fieldType":{"tp":8,"flag":0,"flen":20,"decimal":0,"collate":-63,"charset":"binary"},"hasDistinct":false},"desc":false},{"expr":{"tp":"ColumnRef","val":"gAAAAAAAAAE=","sig":"Unspecified","fieldType":{"tp":15,"flag":0,"flen":100,"decimal":0,"collate":-46,"charset":"utf8mb4"},"hasDistinct":false},"desc":false},{"expr":{"tp":"ColumnRef","val":"gAAAAAAAAAI=","sig":"Unspecified","fieldType":{"tp":8,"flag":0,"flen":20,"decimal":0,"collate":-63,"charset":"binary"},"hasDistinct":false},"desc":false},{"expr":{"tp":"ColumnRef","val":"gAAAAAAAAAA=","sig":"Unspecified","fieldType":{"tp":8,"flag":0,"flen":20,"decimal":0,"collate":-63,"charset":"binary"},"hasDistinct":false},"desc":false}],"isPartialSort":true,"child":{"tp":"TypeExchangeReceiver","exchangeReceiver":{"encodedTaskMeta":["CIGAkPqbr7iCBhABIg8xMjcuMC4wLjE6NDA3Njk="],"fieldTypes":[{"tp":8,"flag":0,"flen":20,"decimal":0,"collate":-63,"charset":"binary"},{"tp":15,"flag":0,"flen":100,"decimal":0,"collate":-46,"charset":"utf8mb4"},{"tp":8,"flag":0,"flen":20,"decimal":0,"collate":-63,"charset":"binary"}]},"executorId":"ExchangeReceiver_16"}})";
```
[the remark this answers is not in the document]
Comment (guo-shaoge) [reply dbms/src/Flash/tests/bench_window.cpp:23]: Good idea, i need take a look.

[the remark this answers is not in the document]
Comment (guo-shaoge) [reply dbms/src/Flash/tests/bench_window.cpp]: Done. Use `DAGRequestBuilder` to build tipb::Window and tipb::Sort.

**Review by SeaRise — APPROVED** [review]

**Review by ywqzzy — APPROVED** [review]

**Pull request merged**